### ISSUE TITLE
Expose maxEvictRetries and machineDrainTimeout to shoot-resource from MCM.

### ISSUE THREAD

**Issue by hardikdr** [gardener/gardener#1843]
**What would you like to be added**:
Currently, MCM exposes `maxEvictRetries` and `machineDrainTimeout` as part of the flags, [refer](https://github.com/gardener/machine-controller-manager/blob/master/cmd/machine-controller-manager/app/options/options.go#L106-L107).  It would be nice to have this exposed as part of the shoot-resource, maybe under `worker` section or separate MCM configuration. 

**Why is this needed**:
At the moment, we set one default value for all the shoots via MCM, this fits for few doesn't for many. Based on the nature of the workload and machines, I think it makes sense to allow setting up the different drain-timeout and eviction-retries per shoot.
- We have also seen a request from a consumer to expose the same.

PS: There is also a partially related issue on MCM for the same: [#372](https://github.com/gardener/machine-controller-manager/issues/372)

**Comment by rfranzke** [comment]
We probably should come up with something similar to https://github.com/gardener/gardener/blob/master/example/90-shoot.yaml#L13, `workerConfig` or so. So far, there is no contract between Gardener and Extensions that MCM is used for managing the machines. Theoretically, it could be anything else. Consequently, directly exposing these flags would be wrong, instead, the providers should define their own worker config API where these MCM-specified flags can be passed. WDYT?

**Comment by hardikdr** [comment]
Yes, this is a very good point, considering the extensions mechanism, MCM specific information should not be directly exposed to shoot. We can then introduce `workerConfig` in parallel and populate the MCM/worker-extension related info there. 

Though, I am curious about the following case. At the moment, both the eviction and drain-timeouts are part of the MCM-flags, they could also be part of the machine classes in the future, considering users willing to set different values for different worker-pools. 
- Then flags could also become part of the same `workerConfig` section in a shoot, right? - and we don't have to touch the `worker`-section as such.

**Comment by rfranzke** [comment]
Hm, no, we already have `providerConfig` in the `.spec.provider.workers[]` list (per pool). It should go there. So, if you plan to make it configurable per worker-pool already then I suggest to move it there directly.

**Comment by dguendisch** [comment]
Would love to see this feature as well. We are running a lot of e2e tests in our testmachinery cluster and when we update the worker pool we'd prefer to let the tests gracefully terminate to avoid repeating a whole testsuite which can take several hours. Hence in our case we'd probably configure the machine-drain-timeout to something like 4 hours (we can live with the rolling taking a whole day or even more but want to avoid breaking tests)

**Comment by rfranzke** [comment]
Currently, these flags are command line parameters of the MCM which is not optimal as end-users might want to tweak them on worker pool basis.
Still, we discussed with @hardikdr and @prashanth26 that there are three options for the implementation:

1. Expose these global flags in a new `.spec.controlPlane.machineControllerManager` section in the shoot.
2. Expose these flags on worker pool basis in the `.spec.provider.workers[].providerConfig` section.
3. Expose these flags on worker pool basis in a new `.spec.provider.workers[].settings` section.

1) has two downsides, firstly it doesn't allow tweaking the parameters on worker pool basis, secondly it adds the machine-controller-manager as required component to the `Worker` extension contract which isn't the case today (though, it's not a strong argument against it as MCM is a first-class citizen of the Gardener project, hence, having a strong dependency against it would be fair).

2) has the downside that every extension controller that is using MCM (i.e., effectively, every existing provider controller today) has to support these configuration flags individually, i.e., more implementation effort + more "duplicate" code

3) has the only downside mentioned in 1) (adding a hard-dependency against MCM). As already mentioned, this is not really a rationale argument given the Gardener project structure. Also, as MCM is currently reworked to support out-of-tree, we might switch back to where Gardener deploys MCM always and the provider extension only adds its provider-specific `Machine` controller (cc @prashanth26) --> In this case we'll have the "hard dependency" anyways.
This approach is the most comfortable wrt implementation effort and robustness against future changes/improvements.

Hence, 3 is preferred over 2), and 2) is preferred over 1), but both 3) and 2) require adaptations in MCM first.

/assign @hardikdr @prashanth26 @rfranzke

**Comment by hardikdr** [comment]
I have opened a [draft PR](https://github.com/gardener/machine-controller-manager/blob/d17eacb313dc0e123987d27b917161f79768c994/kubernetes/machine_objects/machine-deployment.yaml#L30-L35) to expose couple of parameters to MachineDeployment, just to unblock the efforts here meanwhile.

**Comment by rfranzke** [comment]
Thanks. Who will work on the adaptations in g/g?

**Comment by hardikdr** [comment]
I'll pick it up once done with the MCM part.
